### ISSUE TITLE
no-inferrable-types can recommend uncompilable code

### ISSUE THREAD

**Issue by gmoothart** [palantir/tslint#2306]
### Bug Report

- __TSLint version__: 4.4.2
- __TypeScript version__: 2.1
- __Running TSLint via__: CLI

Details on the problem in https://github.com/Microsoft/TypeScript/issues/14416

#### TypeScript code being linted

```ts
export class Hello {
    readonly foo: boolean = false;
    constructor(foo: boolean|undefined) {
        this.foo = foo || this.foo;
    }
}
```

with `tslint.json` configuration:

```json
{
  "rules": {
    "no-inferrable-types": true
  }}
```

#### Actual behavior

tslint error: `Type boolean trivially inferred from a boolean literal, remove type annotation`. But this would cause a compile error.

#### Expected behavior

There should be no error, since the annotation is required in this case.

**Comment by gmoothart** [comment]
@ajafff FYI
